File: python/keyword.py
# =============================================================================
# vapor-idx — indexes/keyword.py
# Tokenised inverted index: "field:token" → set[record_id]
# Supports: contains (field-scoped), free-text search
# All query tokens are AND-ed.
# =============================================================================

from __future__ import annotations
import re
from collections import defaultdict
from typing import Union


class KeywordIndex:
    def __init__(self) -> None:
        # "field:token" → set[id]
        self._index: dict[str, set[str]] = defaultdict(set)
        # id → set["field:token"] for fast removal
        self._record_keys: dict[str, set[str]] = defaultdict(set)

    # ── Mutation ───────────────────────────────────────────────────────────────

    def add(self, field: str, value: object, record_id: str) -> None:
        if value is None:
            return
        raw_values = value if isinstance(value, list) else [value]
        tokens = [t for v in raw_values for t in tokenise(str(v))]
        if not tokens:
            return
        for token in tokens:
            composite_key = f"{field}:{token}"
            self._index[composite_key].add(record_id)
            self._record_keys[record_id].add(composite_key)

    def remove(self, record_id: str) -> None:
        keys = self._record_keys.pop(record_id, set())
        for key in keys:
            id_set = self._index.get(key)
            if id_set:
                id_set.discard(record_id)
                if not id_set:
                    del self._index[key]

    # ── Lookup ─────────────────────────────────────────────────────────────────

    def search(self, query: Union[str, list[str]]) -> set[str]:
        """Free-text search across all keyword-indexed fields. All tokens AND-ed."""
        queries = query if isinstance(query, list) else [query]
        tokens  = [t for q in queries for t in tokenise(q)]
        if not tokens:
            return set()

        per_token: list[set[str]] = []
        for token in tokens:
            merged: set[str] = set()
            for key, ids in self._index.items():
                if key.endswith(f":{token}"):
                    merged.update(ids)
            per_token.append(merged)

        return _intersect_all(per_token)

    def contains(self, field: str, query: Union[str, list[str]]) -> set[str]:
        """Field-scoped search. All tokens must appear in the given field."""
        queries = query if isinstance(query, list) else [query]
        tokens  = [t for q in queries for t in tokenise(q)]
        if not tokens:
            return set()

        per_token = [
            set(self._index.get(f"{field}:{token}", set()))
            for token in tokens
        ]
        return _intersect_all(per_token)

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    def clear(self) -> None:
        self._index.clear()
        self._record_keys.clear()

    @property
    def token_count(self) -> int:
        return len(self._index)


# ── Helpers ───────────────────────────────────────────────────────────────────

_SPLIT_RE = re.compile(r"[\s\-_.,;:!?'\"()\[\]{}<>/\\|@#$%^&*+=~`]+")


def tokenise(value: str) -> list[str]:
    return [t for t in _SPLIT_RE.split(value.lower()) if len(t) > 2]


def _intersect_all(sets: list[set[str]]) -> set[str]:
    if not sets:
        return set()
    sets.sort(key=len)
    result = set(sets[0])
    for s in sets[1:]:
        result.intersection_update(s)
        if not result:
            break
    return result

File: rust/src/indexes/keyword.rs
// =============================================================================
// vapor-idx — indexes/keyword.rs
// Tokenised inverted index.
// =============================================================================

use std::collections::{HashMap, HashSet};

#[derive(Debug, Default)]
pub struct KeywordIndex {
    // "field:token" → set<id>
    index:       HashMap<String, HashSet<String>>,
    // id → set<"field:token"> for O(degree) removal
    record_keys: HashMap<String, HashSet<String>>,
}

impl KeywordIndex {
    pub fn new() -> Self { Self::default() }

    pub fn add(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        let texts: Vec<String> = match value {
            serde_json::Value::String(s) => vec![s.clone()],
            serde_json::Value::Array(arr) => arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect(),
            _ => return,
        };

        let tokens: Vec<String> = texts.iter().flat_map(|t| tokenise(t)).collect();
        if tokens.is_empty() { return; }

        let key_set = self.record_keys.entry(id.to_owned()).or_default();
        for token in tokens {
            let composite = format!("{field}:{token}");
            self.index.entry(composite.clone()).or_default().insert(id.to_owned());
            key_set.insert(composite);
        }
    }

    pub fn remove(&mut self, id: &str) {
        if let Some(keys) = self.record_keys.remove(id) {
            for key in keys {
                if let Some(id_set) = self.index.get_mut(&key) {
                    id_set.remove(id);
                    if id_set.is_empty() { self.index.remove(&key); }
                }
            }
        }
    }

    pub fn search(&self, query: &[String]) -> HashSet<String> {
        let tokens: Vec<String> = query.iter().flat_map(|q| tokenise(q)).collect();
        if tokens.is_empty() { return HashSet::new(); }

        let per_token: Vec<HashSet<String>> = tokens.iter().map(|token| {
            let mut merged = HashSet::new();
            for (key, ids) in &self.index {
                if key.ends_with(&format!(":{token}")) {
                    merged.extend(ids.iter().cloned());
                }
            }
            merged
        }).collect();

        intersect_all(per_token)
    }

    pub fn contains(&self, field: &str, query: &[String]) -> HashSet<String> {
        let tokens: Vec<String> = query.iter().flat_map(|q| tokenise(q)).collect();
        if tokens.is_empty() { return HashSet::new(); }

        let per_token: Vec<HashSet<String>> = tokens.iter().map(|token| {
            let key = format!("{field}:{token}");
            self.index.get(&key).cloned().unwrap_or_default()
        }).collect();

        intersect_all(per_token)
    }

    pub fn clear(&mut self) { self.index.clear(); self.record_keys.clear(); }
    pub fn token_count(&self) -> usize { self.index.len() }
}

pub fn tokenise(value: &str) -> Vec<String> {
    value.to_lowercase()
         .split(|c: char| !c.is_alphanumeric())
         .filter(|t| t.len() > 2)
         .map(String::from)
         .collect()
}

fn intersect_all(mut sets: Vec<HashSet<String>>) -> HashSet<String> {
    if sets.is_empty() { return HashSet::new(); }
    sets.sort_by_key(|s| s.len());
    let mut result = sets[0].clone();
    for s in &sets[1..] {
        result.retain(|id| s.contains(id));
        if result.is_empty() { break; }
    }
    result
}
